File: backend/app/search/rerank.py
from __future__ import annotations

import asyncio
import json
import logging
import os
from typing import Any, Mapping, Protocol, Sequence

import httpx

from app.config import get_settings

logger = logging.getLogger(__name__)

DEFAULT_OPENAI_BASE_URL = "https://api.openai.com/v1"
DEFAULT_RERANK_MODEL = "gpt-4o-mini"
DEFAULT_TIMEOUT_SECONDS = 15.0


class RerankerBackend(Protocol):
    """Pointwise backend: scores one candidate at a time."""

    async def score_relevance(
        self,
        query: str,
        candidate: Mapping[str, Any],
    ) -> float: ...


class BatchRerankerBackend(Protocol):
    """Batch backend: scores all candidates in a single API call.

    Preferred over RerankerBackend when available — lower latency,
    fewer API round-trips, and typically better quality (listwise context).
    """

    async def score_batch(
        self,
        query: str,
        candidates: Sequence[Mapping[str, Any]],
    ) -> list[float]: ...


class OpenAICompatibleRerankerBackend:
    """Pointwise reranker using any OpenAI-compatible chat completions API.

    Makes one LLM call per candidate — suitable for small top_n values.
    """

    def __init__(
        self,
        *,
        api_key: str | None = None,
        base_url: str = DEFAULT_OPENAI_BASE_URL,
        model_name: str = DEFAULT_RERANK_MODEL,
        timeout_seconds: float = DEFAULT_TIMEOUT_SECONDS,
        prompt_template: str | None = None,
    ) -> None:
        settings = get_settings()
        self.api_key = (api_key or os.getenv("OPENAI_API_KEY", "")).strip()
        self.base_url = base_url.rstrip("/")
        self.model_name = model_name
        self.timeout_seconds = timeout_seconds
        self.prompt_template = prompt_template or settings.knowledge.rerank_prompt_template

    async def score_relevance(
        self,
        query: str,
        candidate: Mapping[str, Any],
    ) -> float:
        if not self.api_key:
            raise RuntimeError("OPENAI_API_KEY is not set.")

        prompt = build_rerank_prompt(
            query=query,
            candidate=candidate,
            template_name=self.prompt_template,
        )
        payload = {
            "model": self.model_name,
            "temperature": 0,
            "response_format": {"type": "json_object"},
            "messages": [
                {
                    "role": "system",
                    "content": (
                        "You score how relevant a video segment is to a search query. "
                        "Return JSON with a single numeric field named score from 0 to 10."
                    ),
                },
                {
                    "role": "user",
                    "content": prompt,
                },
            ],
        }

        async with httpx.AsyncClient(timeout=self.timeout_seconds) as client:
            response = await client.post(
                f"{self.base_url}/chat/completions",
                headers={
                    "Authorization": f"Bearer {self.api_key}",
                    "Content-Type": "application/json",
                },
                json=payload,
            )

        response.raise_for_status()
        response_payload = response.json()
        content = _extract_message_content(response_payload)
        parsed_content = json.loads(content)
        if not isinstance(parsed_content, dict) or "score" not in parsed_content:
            raise ValueError("Reranker response did not include a score field.")

        return _clamp_llm_score(float(parsed_content["score"]))


class LLMReranker:
    """Orchestrates reranking using either a pointwise or injected batch backend.

    Automatically uses score_batch() when the backend supports it
    and falls back to parallel score_relevance() calls for pointwise backends.
    """

    def __init__(
        self,
        *,
        backend: RerankerBackend | BatchRerankerBackend | None = None,
        top_n: int | None = None,
    ) -> None:
        settings = get_settings()
        self.backend = backend or _build_default_backend(settings)
        self.top_n = top_n or settings.knowledge.rerank_top_n

    async def rerank(
        self,
        query: str,
        candidates: Sequence[Mapping[str, Any]],
        top_n: int | None = None,
    ) -> list[dict[str, Any]]:
        if not candidates:
            return []

        candidate_limit = min(top_n or self.top_n, len(candidates))
        rerank_candidates = [dict(candidate) for candidate in candidates[:candidate_limit]]
        remaining_candidates = [dict(candidate) for candidate in candidates[candidate_limit:]]

        try:
            if hasattr(self.backend, "score_batch"):
                # Batch backend: one API call for all candidates.
                llm_scores = await self.backend.score_batch(query, rerank_candidates)
                # Batch scores are already 0–1; normalise to match pointwise range.
                llm_scores = [s * 10.0 for s in llm_scores]
            else:
                # Pointwise backend: parallel calls, one per candidate.
                llm_scores = list(
                    await asyncio.gather(
                        *[
                            self.backend.score_relevance(query, candidate)
                            for candidate in rerank_candidates
                        ]
                    )
                )
        except Exception as exc:
            logger.warning(
                "Knowledge reranking failed; falling back to vector score ordering: %s",
                exc,
            )
            return sorted(
                [dict(candidate) for candidate in candidates],
                key=lambda candidate: float(candidate.get("score", 0.0)),
                reverse=True,
            )

        scored_candidates: list[tuple[dict[str, Any], float, int]] = []
        for index, (candidate, llm_score) in enumerate(zip(rerank_candidates, llm_scores)):
            normalized_score = max(0.0, min(llm_score / 10.0, 1.0))
            candidate["llm_score"] = llm_score
            candidate["rerank_score"] = normalized_score
            scored_candidates.append((candidate, normalized_score, index))

        scored_candidates.sort(key=lambda item: (-item[1], item[2]))
        ordered_candidates = [candidate for candidate, _, _ in scored_candidates]
        return ordered_candidates + remaining_candidates


def build_rerank_prompt(
    *,
    query: str,
    candidate: Mapping[str, Any],
    template_name: str = "default",
) -> str:
    transcript_text = _truncate_text(
        _coerce_text(candidate.get("transcript_text") or candidate.get("description")),
        limit=2500,
    )
    visual_description = _truncate_text(
        _coerce_text(
            candidate.get("visual_text_content")
            or candidate.get("visual_description")
            or candidate.get("visual_summary")
            or candidate.get("description")
        ),
        limit=1000,
    )
    visual_type = _coerce_text(candidate.get("visual_type")) or "unknown"
    video_title = _coerce_text(candidate.get("title"))
    speaker = _coerce_text(candidate.get("speaker")) or "Unknown speaker"
    segment_title = _coerce_text(candidate.get("segment_title"))

    if template_name != "default":
        logger.debug("Unknown rerank prompt template '%s'; using default.", template_name)

    return (
        "Search query:\n"
        f"{query}\n\n"
        "Candidate segment:\n"
        f"Video title: {video_title or 'Untitled video'}\n"
        f"Segment title: {segment_title or 'Untitled segment'}\n"
        f"Speaker: {speaker}\n"
        f"Transcript:\n{transcript_text or 'N/A'}\n\n"
        f"Visual type: {visual_type}\n"
        f"Visual evidence:\n{visual_description or 'N/A'}\n\n"
        "Score how useful this segment is for answering the search query.\n"
        "Use 0 for irrelevant and 10 for highly relevant evidence.\n"
        'Return JSON only, for example: {"score": 8.5}.'
    )


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _build_default_backend(
    settings: Any,
) -> RerankerBackend:
    model = getattr(settings.knowledge, "rerank_model", DEFAULT_RERANK_MODEL)
    return OpenAICompatibleRerankerBackend(model_name=model)


def _extract_message_content(payload: Mapping[str, Any]) -> str:
    choices = payload.get("choices")
    if not isinstance(choices, list) or not choices:
        raise ValueError("LLM response did not include choices.")

    first_choice = choices[0]
    if not isinstance(first_choice, Mapping):
        raise ValueError("LLM choice payload is malformed.")

    message = first_choice.get("message")
    if not isinstance(message, Mapping):
        raise ValueError("LLM response did not include a message.")

    content = message.get("content")
    if isinstance(content, str):
        return content

    if isinstance(content, list):
        text_fragments = []
        for item in content:
            if isinstance(item, Mapping) and isinstance(item.get("text"), str):
                text_fragments.append(item["text"])
        joined_content = "".join(text_fragments).strip()
        if joined_content:
            return joined_content

    raise ValueError("LLM response did not include message content.")


def _clamp_llm_score(score: float) -> float:
    return max(0.0, min(score, 10.0))


def _coerce_text(value: Any) -> str:
    if value is None:
        return ""
    return str(value).strip()


def _truncate_text(value: str, *, limit: int) -> str:
    if len(value) <= limit:
        return value
    return value[: limit - 3].rstrip() + "..."

File: api/src/services/rerank.ts
import type { AppConfig, Bindings } from "../types";

const DEFAULT_OPENAI_BASE_URL = "https://api.openai.com/v1";
const DEFAULT_RERANK_MODEL = "jina-reranker-v3";
const DEFAULT_JINA_BASE_URL = "https://api.jina.ai/v1";

function clampScore(score: number): number {
  return Math.max(0, Math.min(score, 10));
}

function coerceText(value: unknown): string {
  return value == null ? "" : String(value).trim();
}

function truncateText(value: string, limit: number): string {
  return value.length <= limit ? value : `${value.slice(0, limit - 3).trimEnd()}...`;
}

function buildJinaDocument(candidate: Record<string, unknown>): string {
  const parts: string[] = [];
  const title = coerceText(candidate.title);
  if (title) {
    parts.push(title);
  }
  const segmentTitle = coerceText(candidate.segment_title);
  if (segmentTitle && segmentTitle !== title) {
    parts.push(segmentTitle);
  }
  const transcript = truncateText(
    coerceText(candidate.transcript_text ?? candidate.description),
    1500
  );
  if (transcript) {
    parts.push(transcript);
  }
  const visual = truncateText(
    coerceText(candidate.visual_text_content ?? candidate.visual_description ?? candidate.visual_summary),
    500
  );
  if (visual) {
    parts.push(visual);
  }
  return parts.length > 0 ? parts.join("\n") : "N/A";
}

function buildRerankPrompt(query: string, candidate: Record<string, unknown>, templateName = "default"): string {
  if (templateName !== "default") {
    // Keep parity with Python behavior: unknown templates silently fall back.
  }
  const transcriptText = truncateText(coerceText(candidate.transcript_text ?? candidate.description), 2500);
  const visualDescription = truncateText(
    coerceText(
      candidate.visual_text_content ??
      candidate.visual_description ??
      candidate.visual_summary ??
      candidate.description
    ),
    1000
  );
  const visualType = coerceText(candidate.visual_type) || "unknown";
  const videoTitle = coerceText(candidate.title);
  const speaker = coerceText(candidate.speaker) || "Unknown speaker";
  const segmentTitle = coerceText(candidate.segment_title);

  return (
    "Search query:\n" +
    `${query}\n\n` +
    "Candidate segment:\n" +
    `Video title: ${videoTitle || "Untitled video"}\n` +
    `Segment title: ${segmentTitle || "Untitled segment"}\n` +
    `Speaker: ${speaker}\n` +
    `Transcript:\n${transcriptText || "N/A"}\n\n` +
    `Visual type: ${visualType}\n` +
    `Visual evidence:\n${visualDescription || "N/A"}\n\n` +
    "Score how useful this segment is for answering the search query.\n" +
    "Use 0 for irrelevant and 10 for highly relevant evidence.\n" +
    'Return JSON only, for example: {"score": 8.5}.'
  );
}

function extractMessageContent(payload: any): string {
  const content = payload?.choices?.[0]?.message?.content;
  if (typeof content === "string") {
    return content;
  }
  if (Array.isArray(content)) {
    return content.map((item) => (typeof item?.text === "string" ? item.text : "")).join("").trim();
  }
  throw new Error("LLM response did not include message content.");
}

async function scoreWithJina(env: Bindings, modelName: string, query: string, candidates: Array<Record<string, unknown>>): Promise<number[]> {
  const apiKey = (env.JINA_API_KEY ?? "").trim();
  if (!apiKey) {
    throw new Error("JINA_API_KEY is not set.");
  }

  const response = await fetch(`${DEFAULT_JINA_BASE_URL}/rerank`, {
    method: "POST",
    headers: {
      authorization: `Bearer ${apiKey}`,
      "content-type": "application/json"
    },
    body: JSON.stringify({
      model: modelName,
      query,
      documents: candidates.map(buildJinaDocument),
      top_n: candidates.length,
      truncation: true,
      return_documents: false
    })
  });
  if (!response.ok) {
    throw new Error(`Jina rerank request failed: ${response.status}`);
  }

  const payload = await response.json();
  const scores = new Array<number>(candidates.length).fill(0);
  for (const result of Array.isArray(payload?.results) ? payload.results : []) {
    const index = Number(result?.index ?? 0);
    if (index >= 0 && index < scores.length) {
      const relevanceScore = Number(result?.relevance_score ?? 0);
      scores[index] = Math.max(0, Math.min(relevanceScore, 1)) * 10;
    }
  }
  return scores;
}

async function scoreWithOpenAI(
  env: Bindings,
  config: AppConfig,
  modelName: string,
  query: string,
  candidates: Array<Record<string, unknown>>
): Promise<number[]> {
  const apiKey = (env.OPENAI_API_KEY ?? "").trim();
  if (!apiKey) {
    throw new Error("OPENAI_API_KEY is not set.");
  }

  const baseUrl = DEFAULT_OPENAI_BASE_URL;
  const responses = await Promise.all(
    candidates.map(async (candidate) => {
      const response = await fetch(`${baseUrl}/chat/completions`, {
        method: "POST",
        headers: {
          authorization: `Bearer ${apiKey}`,
          "content-type": "application/json"
        },
        body: JSON.stringify({
          model: modelName,
          temperature: 0,
          response_format: { type: "json_object" },
          messages: [
            {
              role: "system",
              content: "You score how relevant a video segment is to a search query. Return JSON with a single numeric field named score from 0 to 10."
            },
            {
              role: "user",
              content: buildRerankPrompt(query, candidate, config.knowledge.rerankPromptTemplate)
            }
          ]
        })
      });
      if (!response.ok) {
        throw new Error(`OpenAI rerank request failed: ${response.status}`);
      }
      const payload = await response.json();
      const content = extractMessageContent(payload);
      const parsed = JSON.parse(content);
      return clampScore(Number(parsed.score ?? 0));
    })
  );

  return responses;
}

export class LLMReranker {
  private readonly topN: number;

  constructor(
    private readonly env: Bindings,
    private readonly config: AppConfig,
    topN?: number
  ) {
    this.topN = topN ?? config.knowledge.rerankTopN;
  }

  async rerank(query: string, candidates: Array<Record<string, unknown>>, topN?: number): Promise<Array<Record<string, unknown>>> {
    if (candidates.length === 0) {
      return [];
    }

    const candidateLimit = Math.min(topN ?? this.topN, candidates.length);
    const rerankCandidates = candidates.slice(0, candidateLimit).map((candidate) => ({ ...candidate }));
    const remainingCandidates = candidates.slice(candidateLimit).map((candidate) => ({ ...candidate }));
    const modelName = this.config.knowledge.rerankModel || DEFAULT_RERANK_MODEL;

    let scores: number[];
    try {
      scores = modelName.toLowerCase().includes("jina")
        ? await scoreWithJina(this.env, modelName, query, rerankCandidates)
        : await scoreWithOpenAI(this.env, this.config, modelName, query, rerankCandidates);
    } catch {
      return [...candidates].sort((left, right) => Number(right.score ?? 0) - Number(left.score ?? 0));
    }

    const scoredCandidates = rerankCandidates.map((candidate, index) => {
      const rerankScore = Math.max(0, Math.min(scores[index] / 10, 1));
      return {
        candidate: {
          ...candidate,
          llm_score: scores[index],
          rerank_score: rerankScore
        },
        rerankScore,
        index
      };
    });

    scoredCandidates.sort((left, right) => {
      if (right.rerankScore !== left.rerankScore) {
        return right.rerankScore - left.rerankScore;
      }
      return left.index - right.index;
    });

    return [...scoredCandidates.map((item) => item.candidate), ...remainingCandidates];
  }
}
